### ISSUE TITLE
unparse (round-tripping)

### ISSUE THREAD

**Issue by goretkin** [julia-vscode/CSTParser.jl#265]
Is it in the spirit of this package to provide a way to turn the CST back into text? What kind of round-tripping is possible or intended to be possible?

**Comment by davidanthoff** [comment]
Yes, that is indeed one of the main points of this package :) Full lossless round tripping. Now, how to do that, you'll have to ask @ZacLN :)

**Comment by goretkin** [comment]
These two seem to produce the exact same CST. Is there a way to distinguish these?
```
CSTParser.parse("a+ b", true)
CSTParser.parse("a+\nb", true)
```

checked with

```julia
julia> tempfileshim(dump, r1) = let (fn, io) = mktemp(); dump(io, r1); close(io); fn; end
tempfileshim (generic function with 1 method)

shell> diff $(tempfileshim(dump, CSTParser.parse("a+ b", true))) $(tempfileshim(dump, CSTParser.parse("a+\nb", true)))
```

**Comment by pfitzseb** [comment]
Pretty sure you need to refer back to the parsed string for that (in particular, to the part between the `+`s span and fullspan).

**Comment by goretkin** [comment]
@pfitzseb if you need to refer back to the string, then the CST is not a complete representation, which seems to go against the main point as mentioned by @davidanthoff .

**Comment by goretkin** [comment]
where `x::EXPR`, in addition to a  `x.val` field, could there also be a `x.val_whitespace` field?

These fields are `String`s (or `nothing`) of length `x.span` and ``x.fullspan - x.span` respectively.

This field would serve the role of `trailing_ws` in the code below:

```julia
function flatten(io::IO, cst)
    if isempty(cst)
        print(io, cst.val)
        length(cst.val) == cst.span || error("GNG misunderstood")
        # TODO, should contain kind of white space (\n, \r, \t, ...)
        trailing_ws = " "^(cst.fullspan - cst.span)
        print(io, trailing_ws)
    else
        cst.val === nothing || error("GNG misunderstood")
    end
    for arg_or_trivia in cst
        flatten(io, arg_or_trivia)
    end
end
```

**Comment by goretkin** [comment]
I suppose it's actually more than just trailing whitespace that is missing from the `EXPR` representation, e.g.:

`CSTParser.parse("a + #=interruption=# b", true)`

**Comment by ZacLN** [comment]
Round-tripping was initially the goal but that was dropped (with respect to whitespace/comments) as there wasn't enough of a benefit from that approach to outweigh the performance costs. For our uses (julia-vscode, etc) we always have the parsed string alongside the CST (and access it for ws/comments as per @pfitzseb 's comment above).

**Comment by davidanthoff** [comment]
Ha, I completely missed that we no longer support that :)

**Comment by pfitzseb** [comment]
Btw, I'm pretty sure that JuliaFormatter's `FST` includes all necessary info.

**Comment by goretkin** [comment]
@pfitzseb thanks. I'll have to take another look, but I was under the [opposite impression](https://github.com/domluna/JuliaFormatter.jl/blob/7d2c619fa45ac7d89e864eac24ba455e781a9203/docs/src/how_it_works.md):


> The important part of an FST is any `.jl` file that is syntactically the same (whitespace is irrelevant) produce an identical
`FST`.

@domluna , could you clarify that both a `JuliaFormatter.FST` and a `JuliaFormater.Document` are needed to round-trip back to the source text losslessly?

**Comment by domluna** [comment]
Yeah you would need the Document as well which contains the original source. The FST isn't meant to be able to go back to the original source since it runs counter to its purpose. 

That being said I think FST has all the info to actually do this since nodes have the original line offset (used for alignment) and start/end lines (used to splice comments) but there's currently no utility to help with this. The point of keeping source information w.r.t. the FST is that for certain formatting transformations we want to maintain how things were in the original source.

**Comment by davidanthoff** [comment]
There is no todo here, right? So I'm going to close this.
